### PR TITLE
Refactor Kernel

### PR DESCRIPTION
**Pull request by blink1073**
cf #190.

Moves kernel-related files and code to `src/kernel`.

API changes: 

- Use a `Kernel` namespace instead of the `IKernel` namespace.
- Rename `Kernel` class -> `DefaultKernel`.  
- Move the free kernel-related functions onto the `Kernel` and `KernelMessage` namespaces as appropriate.
- Kernel now exports: `IKernel` interface, `Kernel` namespace, `KernelMessage` namespace, `KernelManager` class. 
- Rename `IKernel.getKernelSpecs` -> `IKernel.getSpecs`

### CODE REVIEW

File: src/kernel/default.ts [hunk 1]
```diff
@@ -1267,231 +916,270 @@ class Kernel implements IKernel {
...
+namespace DefaultKernel {
   /**
-   * Construct a new comm channel.
+   * Find a kernel by id.
+   *
+   * #### Notes
+   * If the kernel was already started via `startNewKernel`, we return its
+   * `Kernel.IModel`.
+   *
+   * Otherwise, if `options` are given, we attempt to find to the existing
```
Comment (afshin) [inline src/kernel/default.ts]: >  to find ~~to~~ the

File: src/kernel/index.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+'use strict';
```
Comment (afshin) [inline src/kernel/index.ts]: Superfluous `use strict`.

File: src/kernel/kernel.ts [hunk 3]
```diff
@@ -0,0 +1,738 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+'use strict';
```
Comment (afshin) [inline src/kernel/kernel.ts]: Superfluous `use strict`.

File: src/kernel/index.ts (hunk 2 above)
Comment (afshin) [inline src/kernel/index.ts]: There are a bunch of these, I won't bombard the PR marking all of them.

File: src/kernel/default.ts [hunk 4]
```diff
@@ -610,7 +259,7 @@ class Kernel implements IKernel {
    *
    * If the kernel status is `Dead`, this will throw an error.
```
Comment (afshin) [inline src/kernel/default.ts]: In the doc strings it says `Dead` but in the code it's `dead` ... I don't know if this makes a big difference, but we may want to sync those up.

File: src/kernel/index.ts (hunk 2 above)
Comment (blink1073) [inline src/kernel/index.ts]: I updated all of the ones in the `kernel` folders.

**Comment by blink1073** [comment]
Comments addressed.

**Comment by afshin** [comment]
Awesome, looks good to me.

👍

**Pull request merged**